// bitfab-plugin-lib/dist/commands/label.js
import { openChromelessWindow } from "../browser.js";
import { getConfig, hasCredentials } from "../config.js";
import { startLoopbackServer } from "../loopback.js";
import { findOpenPort, focusApp, getFrontmostApp } from "./login.js";
const LABEL_TIMEOUT_MS = 30 * 60 * 1000;
const RESPONSE_HTML = `<html>
<body style="margin:0;background:#f8fafc;color:#0f172a;font-family:system-ui,-apple-system,sans-serif;display:flex;align-items:center;justify-content:center;height:100vh;overflow:hidden">
  <div style="text-align:center;max-width:400px">
    <h1 style="font-size:24px;font-weight:600;margin:0 0 16px 0">Bitfab</h1>
    <p style="margin:0;color:#059669">Labeling complete! You can close this window.</p>
    <button onclick="window.close()" style="margin-top:16px;padding:8px 16px;background:#0f172a;color:white;border:none;border-radius:8px;font-size:14px;font-weight:500;cursor:pointer">Close Window</button>
  </div>
</body>
</html>`;
export async function runLabel(platform) {
    const args = process.argv.slice(2);
    const functionKey = args[0];
    if (!functionKey) {
        console.error("Usage: label <functionKey>");
        console.error("Function key is required.");
        process.exit(1);
    }
    if (!hasCredentials()) {
        console.error(`Not authenticated. Run ${platform.loginHint} first.`);
        process.exit(1);
    }
    const config = getConfig();
    const previousApp = getFrontmostApp();
    const port = await findOpenPort();
    const loopback = await startLoopbackServer(port, [
        {
            path: "/done",
            handle: (params, res, resolve) => {
                res.writeHead(200, { "Content-Type": "text/html" });
                res.end(RESPONSE_HTML);
                resolve({
                    status: params.get("status") ?? "unknown",
                    decision: params.get("decision"),
                    labeled: params.get("labeled"),
                    total: params.get("total"),
                    isFetch: params.get("isFetch") === "true",
                });
            },
        },
    ]);
    // Labels are saved server-side as the user clicks through — the loopback
    // callback only delivers the final "N/M labeled" summary line and signals
    // the CLI to exit. If loopback can't bind, the labeling itself still works
    // in the browser; we just can't auto-exit. Let the user Ctrl-C when done
    // (the 30-min timeout is the safety net).
    if (!loopback.available) {
        console.log(`Loopback server unavailable (${loopback.bindError?.message ?? "unknown"}). Labeling will still work in the browser — press Ctrl-C here when you're done.`);
    }
    const labelUrl = new URL(`${config.serviceUrl}/plugin/dataset/${encodeURIComponent(functionKey)}`);
    labelUrl.searchParams.set("callbackPort", String(port));
    console.log(`Opening dataset review page: ${labelUrl}`);
    openChromelessWindow(labelUrl.toString());
    console.log("Waiting for labeling to complete...");
    const timeoutTimer = setTimeout(() => {
        loopback.reject(new Error(`Labeling timed out after ${Math.round(LABEL_TIMEOUT_MS / 60_000)} minutes.`));
    }, LABEL_TIMEOUT_MS);
    let isFetch = false;
    try {
        const result = await loopback.promise;
        isFetch = result.isFetch;
        const { status, decision, labeled, total } = result;
        if (status === "saved") {
            if (labeled && total) {
                console.log(`Labeling complete: ${labeled}/${total} traces labeled (last: ${decision ?? "labeled"})`);
            }
            else {
                console.log(`Labeling complete: ${decision ?? "labeled"}`);
            }
        }
        else {
            if (labeled && total && Number(labeled) > 0) {
                console.log(`Labeling ended early: ${labeled}/${total} traces labeled.`);
            }
            else {
                console.log("Labeling ended without saving.");
            }
        }
    }
    catch (err) {
        console.log(err.message);
    }
    finally {
        clearTimeout(timeoutTimer);
    }
    // Brief pause so the browser can render the success HTML response before
    // the loopback server shuts down on process exit.
    setTimeout(() => {
        if (isFetch) {
            focusApp(previousApp);
        }
        loopback.close();
        process.exit(0);
    }, 500);
}
