// bitfab-plugin-lib/dist/commands/waitForTrace.js
import { getConfig, hasCredentials } from "../config.js";
import { McpProxy } from "../mcpProxy.js";
const DEFAULT_POLL_INTERVAL_MS = 10_000;
const DEFAULT_TIMEOUT_MS = 10 * 60 * 1000;
const TRACE_ID_REGEX = /ID:\s+([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})/i;
/**
 * Pure function — extracts the first trace UUID from the plaintext response
 * returned by the server's `search_traces` tool. Returns null when no match
 * exists (e.g. "No traces found matching the filter criteria.").
 */
export function extractFirstTraceId(text) {
    const match = text.match(TRACE_ID_REGEX);
    return match?.[1] ?? null;
}
/**
 * Pure function — parses argv for wait-for-trace. First non-flag arg is the
 * trace function key; `--timeout-ms` / `--poll-interval-ms` override defaults.
 * Throws when the key is missing or a flag argument is malformed.
 */
export function parseWaitForTraceArgs(argv) {
    const traceFunctionKey = argv[0];
    if (!traceFunctionKey || traceFunctionKey.startsWith("--")) {
        throw new Error("Usage: wait-for-trace <traceFunctionKey> [--timeout-ms N] [--poll-interval-ms N]");
    }
    return {
        traceFunctionKey,
        timeoutMs: parseFlagMs(argv, "--timeout-ms", DEFAULT_TIMEOUT_MS),
        pollIntervalMs: parseFlagMs(argv, "--poll-interval-ms", DEFAULT_POLL_INTERVAL_MS),
    };
}
/**
 * Core polling loop — repeatedly calls `search_traces` against the given
 * MCP proxy until the first trace arrives, the timeout elapses, or the
 * caller signals interrupt. Exported so tests can drive it with a fake
 * proxy + config instead of real creds + HTTP.
 */
export async function pollForFirstTrace(opts) {
    const startMs = Date.now();
    const isInterrupted = opts.isInterrupted ?? (() => false);
    while (!isInterrupted() && Date.now() - startMs < opts.timeoutMs) {
        const result = await opts.proxy.toolCall(opts.config, "search_traces", {
            traceFunctionKey: opts.traceFunctionKey,
            limit: 1,
        });
        const text = result.content?.[0]?.text ?? "";
        if (result.isError) {
            return { status: "error", message: text || "search_traces error" };
        }
        const traceId = extractFirstTraceId(text);
        if (traceId) {
            const serviceUrl = opts.config.serviceUrl.replace(/\/$/, "");
            return {
                status: "found",
                traceId,
                url: `${serviceUrl}/traces/${traceId}`,
                waitedMs: Date.now() - startMs,
            };
        }
        await sleep(opts.pollIntervalMs, isInterrupted);
    }
    return {
        status: isInterrupted() ? "interrupted" : "timeout",
        waitedMs: Date.now() - startMs,
    };
}
/**
 * CLI entrypoint. Parses argv, checks creds, invokes `pollForFirstTrace`,
 * prints one JSON line to stdout, and exits. Exists so setup skills can
 * block on the first trace via a single long-timeout `Bash` call instead
 * of a 40-iteration agent-side poll loop.
 */
export async function runWaitForTrace(platform, pluginVersion) {
    let args;
    try {
        args = parseWaitForTraceArgs(process.argv.slice(2));
    }
    catch (err) {
        console.error(err.message);
        process.exit(1);
    }
    if (!hasCredentials()) {
        console.error(`Not authenticated. Run ${platform.loginHint} first.`);
        process.exit(1);
    }
    let interrupted = false;
    const onSignal = () => {
        interrupted = true;
    };
    process.on("SIGINT", onSignal);
    process.on("SIGTERM", onSignal);
    const result = await pollForFirstTrace({
        proxy: new McpProxy(platform, pluginVersion),
        config: getConfig(),
        traceFunctionKey: args.traceFunctionKey,
        timeoutMs: args.timeoutMs,
        pollIntervalMs: args.pollIntervalMs,
        isInterrupted: () => interrupted,
    });
    if (result.status === "error") {
        console.error(result.message);
        process.exit(1);
    }
    console.log(JSON.stringify(result));
    process.exit(0);
}
function parseFlagMs(args, flag, fallback) {
    const idx = args.indexOf(flag);
    if (idx === -1) {
        return fallback;
    }
    const raw = args[idx + 1];
    const parsed = Number(raw);
    return Number.isFinite(parsed) && parsed > 0 ? parsed : fallback;
}
function sleep(ms, isInterrupted) {
    return new Promise((resolve) => {
        const interval = 200;
        let elapsed = 0;
        const timer = setInterval(() => {
            elapsed += interval;
            if (isInterrupted() || elapsed >= ms) {
                clearInterval(timer);
                resolve();
            }
        }, interval);
    });
}
