// bitfab-plugin-lib/dist/browserHandoff.js
import http from "node:http";
import { openChromelessWindow } from "./browser.js";
import { startHandoffTicketChannel } from "./handoffTicketChannel.js";
import { startLoopbackServer } from "./loopback.js";
const DEFAULT_TIMEOUT_MS = 10 * 60 * 1000;
export function findOpenPort() {
    return new Promise((resolve, reject) => {
        const server = http.createServer();
        server.listen(0, () => {
            const addr = server.address();
            if (addr && typeof addr === "object") {
                const port = addr.port;
                server.close(() => resolve(port));
            }
            else {
                reject(new Error("Could not find open port"));
            }
        });
    });
}
/**
 * Low-level primitive: start both channels + the browser, return handles so
 * the caller can compose additional racers (e.g. login's interactive paste)
 * and print their own hint messages. Most callers should use the higher-level
 * `runBrowserHandoff` instead.
 */
export async function startBrowserHandoff(opts) {
    const port = await findOpenPort();
    const timeoutMs = opts.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    // Debug flag: force ticket-only delivery even on a machine where
    // loopback would otherwise bind. Used to exercise the server-polled
    // ticket path locally without needing an SSH/Docker/container env.
    const forceTicket = process.env.BITFAB_FORCE_TICKET === "1";
    const [loopback, ticket] = await Promise.all([
        forceTicket
            ? Promise.resolve({
                promise: new Promise(() => { }),
                get done() {
                    return false;
                },
                available: false,
                bindError: new Error("BITFAB_FORCE_TICKET=1 disabled loopback"),
                resolve: () => { },
                reject: () => { },
                close: () => { },
            })
            : startLoopbackServer(port, opts.loopbackRoutes),
        startHandoffTicketChannel({
            serviceUrl: opts.serviceUrl,
            kind: opts.ticketKind,
            timeoutMs,
            ttlSeconds: opts.ticketTtlSeconds,
            bearer: opts.ticketBearer,
        }),
    ]);
    const browserUrl = opts.buildUrl({
        port: loopback.available ? port : null,
        ticketId: ticket.ticketId,
    });
    return { loopback, ticket, port, browserUrl, timeoutMs };
}
/**
 * High-level flow: open a browser window to `buildUrl(...)` and race the
 * loopback and ticket channels until one delivers a value of type `T`. Tears
 * down both channels in a `finally` block.
 *
 * Use this for any browser → CLI hand-off that doesn't need extras like
 * interactive paste, SSH-specific hints, or short-circuit timeouts — e.g.
 * dataset selection, grader editing, experiment configuration. The login
 * flow uses `startBrowserHandoff` directly to layer on its extras.
 */
export async function runBrowserHandoff(opts) {
    const { loopback, ticket, browserUrl, timeoutMs } = await startBrowserHandoff(opts);
    if (!loopback.available && !ticket.available) {
        loopback.close();
        ticket.close();
        throw new Error(`Neither loopback nor ticket channel is available — can't complete the browser hand-off. Loopback: ${loopback.bindError?.message ?? "ok"}; ticket: ${ticket.startError?.message ?? "ok"}.`);
    }
    console.log("\nOpening your browser...");
    console.log(`  ${browserUrl}`);
    console.log("");
    console.log("(If the browser didn't open, visit the URL above manually.)");
    openChromelessWindow(browserUrl);
    const timeoutPromise = new Promise((_resolve, reject) => {
        setTimeout(() => reject(new Error(`Browser hand-off timed out after ${Math.round(timeoutMs / 1000)}s.`)), timeoutMs);
    });
    const channelAvailability = {
        loopbackAvailable: loopback.available,
        ticketAvailable: ticket.available,
    };
    const racers = [];
    if (loopback.available) {
        racers.push(loopback.promise.then((value) => ({
            value,
            wonBy: "loopback",
            ...channelAvailability,
        })));
    }
    if (ticket.available) {
        racers.push(ticket.promise.then((value) => ({
            value,
            wonBy: "ticket",
            ...channelAvailability,
        })));
    }
    try {
        return await Promise.race([...racers, timeoutPromise]);
    }
    finally {
        loopback.close();
        ticket.close();
    }
}
