// bitfab-plugin-lib/dist/buildInfo.js
import crypto from "node:crypto";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";
export function readBuildInfo(distDir) {
    try {
        const raw = fs.readFileSync(path.join(distDir, "buildInfo.json"), "utf-8");
        const parsed = JSON.parse(raw);
        if (typeof parsed.builtAt === "string" &&
            typeof parsed.srcHash === "string") {
            return {
                gitSha: typeof parsed.gitSha === "string" ? parsed.gitSha : null,
                builtAt: parsed.builtAt,
                srcHash: parsed.srcHash,
            };
        }
        return null;
    }
    catch {
        return null;
    }
}
export function hashSrcDir(srcDir) {
    const hash = crypto.createHash("sha256");
    const walk = (dir) => {
        let entries;
        try {
            entries = fs.readdirSync(dir, { withFileTypes: true });
        }
        catch {
            return;
        }
        const sorted = entries.slice().sort((a, b) => a.name.localeCompare(b.name));
        for (const entry of sorted) {
            const full = path.join(dir, entry.name);
            if (entry.isDirectory()) {
                walk(full);
            }
            else if (entry.isFile() &&
                entry.name.endsWith(".ts") &&
                !entry.name.endsWith(".test.ts")) {
                hash.update(path.relative(srcDir, full));
                hash.update("\0");
                hash.update(fs.readFileSync(full));
                hash.update("\0");
            }
        }
    };
    walk(srcDir);
    return hash.digest("hex").slice(0, 12);
}
export function isDevBuildActive(platform, scriptUrl) {
    const devKey = `${platform.pluginName}@${platform.marketplaceName}-dev`;
    for (const settingsPath of [
        path.join(os.homedir(), ".claude", "settings.json"),
        path.join(os.homedir(), ".claude", "settings.local.json"),
        path.join(process.cwd(), ".claude", "settings.json"),
        path.join(process.cwd(), ".claude", "settings.local.json"),
    ]) {
        try {
            const data = JSON.parse(fs.readFileSync(settingsPath, "utf-8"));
            if (data.enabledPlugins?.[devKey] === true) {
                return true;
            }
        }
        catch {
            // settings file missing or unreadable
        }
    }
    if (scriptUrl) {
        try {
            const scriptPath = fileURLToPath(scriptUrl);
            const cursorLocal = `${path.join(os.homedir(), ".cursor", "plugins", "local")}${path.sep}`;
            if (scriptPath.startsWith(cursorLocal)) {
                return true;
            }
        }
        catch {
            // scriptUrl not a file URL
        }
    }
    return false;
}
export function formatRelativeTime(isoTimestamp, now = Date.now()) {
    const then = new Date(isoTimestamp).getTime();
    if (Number.isNaN(then)) {
        return isoTimestamp;
    }
    const deltaMs = now - then;
    if (deltaMs < 0) {
        return "just now";
    }
    const sec = Math.floor(deltaMs / 1000);
    if (sec < 60) {
        return `${sec}s ago`;
    }
    const min = Math.floor(sec / 60);
    if (min < 60) {
        return `${min}m ago`;
    }
    const hr = Math.floor(min / 60);
    if (hr < 24) {
        return `${hr}h ago`;
    }
    const d = Math.floor(hr / 24);
    return `${d}d ago`;
}
export function getDevBuildSummary(pluginRoot) {
    const distDir = path.join(pluginRoot, "dist");
    const info = readBuildInfo(distDir);
    if (!info) {
        return null;
    }
    const srcDir = path.join(pluginRoot, "src");
    let srcMatches = true;
    try {
        if (fs.statSync(srcDir).isDirectory()) {
            srcMatches = hashSrcDir(srcDir) === info.srcHash;
        }
    }
    catch {
        // No src/ present (vendored install) — nothing to compare against.
        srcMatches = true;
    }
    return {
        info,
        pluginRoot,
        srcMatches,
        relativeBuiltAt: formatRelativeTime(info.builtAt),
    };
}
export function formatDevBuildStatusLine(summary) {
    const sha = summary.info.gitSha ? summary.info.gitSha.slice(0, 7) : "unknown";
    const match = summary.srcMatches
        ? "src matches: ✅"
        : "src matches: ❌ run pnpm build";
    return `Build: ${sha} · ${summary.relativeBuiltAt} · ${match}`;
}
export function formatDevBuildBannerLines(summary, platformLabel) {
    const sha = summary.info.gitSha ? summary.info.gitSha.slice(0, 7) : "unknown";
    const lines = [
        `[${platformLabel} Dev] build ${sha} · ${summary.relativeBuiltAt} · ${summary.pluginRoot}`,
    ];
    if (!summary.srcMatches) {
        lines.push(`             src changed since last build — run pnpm build to apply.`);
    }
    return lines;
}
