// bitfab-plugin-lib/dist/buildInfo.test.js
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { afterEach, beforeEach, describe, expect, it } from "vitest";
import { formatDevBuildBannerLines, formatDevBuildStatusLine, formatRelativeTime, getDevBuildSummary, hashSrcDir, readBuildInfo, } from "./buildInfo.js";
let tmp;
beforeEach(() => {
    tmp = fs.mkdtempSync(path.join(os.tmpdir(), "bitfab-buildinfo-"));
});
afterEach(() => {
    fs.rmSync(tmp, { recursive: true, force: true });
});
describe("hashSrcDir", () => {
    it("returns the same 12-char hash for identical trees", () => {
        const a = path.join(tmp, "a/src");
        const b = path.join(tmp, "b/src");
        fs.mkdirSync(path.join(a, "hooks"), { recursive: true });
        fs.mkdirSync(path.join(b, "hooks"), { recursive: true });
        fs.writeFileSync(path.join(a, "index.ts"), "export const x = 1\n");
        fs.writeFileSync(path.join(b, "index.ts"), "export const x = 1\n");
        fs.writeFileSync(path.join(a, "hooks/h.ts"), "export default null\n");
        fs.writeFileSync(path.join(b, "hooks/h.ts"), "export default null\n");
        const hashA = hashSrcDir(a);
        const hashB = hashSrcDir(b);
        expect(hashA).toBe(hashB);
        expect(hashA).toHaveLength(12);
    });
    it("changes when a source file is modified", () => {
        const src = path.join(tmp, "src");
        fs.mkdirSync(src, { recursive: true });
        fs.writeFileSync(path.join(src, "a.ts"), "export const x = 1\n");
        const before = hashSrcDir(src);
        fs.writeFileSync(path.join(src, "a.ts"), "export const x = 2\n");
        const after = hashSrcDir(src);
        expect(before).not.toBe(after);
    });
    it("ignores .test.ts files", () => {
        const src = path.join(tmp, "src");
        fs.mkdirSync(src, { recursive: true });
        fs.writeFileSync(path.join(src, "a.ts"), "export const x = 1\n");
        const before = hashSrcDir(src);
        fs.writeFileSync(path.join(src, "a.test.ts"), "it('x', () => {})\n");
        const after = hashSrcDir(src);
        expect(before).toBe(after);
    });
});
describe("readBuildInfo", () => {
    it("returns null when file is missing", () => {
        expect(readBuildInfo(tmp)).toBeNull();
    });
    it("returns null for malformed JSON", () => {
        fs.writeFileSync(path.join(tmp, "buildInfo.json"), "not json");
        expect(readBuildInfo(tmp)).toBeNull();
    });
    it("returns null when required fields are missing", () => {
        fs.writeFileSync(path.join(tmp, "buildInfo.json"), JSON.stringify({ gitSha: "abc" }));
        expect(readBuildInfo(tmp)).toBeNull();
    });
    it("parses a valid file", () => {
        const payload = {
            gitSha: "abc1234",
            builtAt: "2026-04-15T00:00:00Z",
            srcHash: "deadbeef0000",
        };
        fs.writeFileSync(path.join(tmp, "buildInfo.json"), JSON.stringify(payload));
        expect(readBuildInfo(tmp)).toEqual(payload);
    });
});
describe("getDevBuildSummary", () => {
    it("returns null when dist/buildInfo.json is absent", () => {
        fs.mkdirSync(path.join(tmp, "dist"), { recursive: true });
        expect(getDevBuildSummary(tmp)).toBeNull();
    });
    it("reports srcMatches=true when src hash matches", () => {
        const src = path.join(tmp, "src");
        const dist = path.join(tmp, "dist");
        fs.mkdirSync(src, { recursive: true });
        fs.mkdirSync(dist, { recursive: true });
        fs.writeFileSync(path.join(src, "a.ts"), "export const x = 1\n");
        const srcHash = hashSrcDir(src);
        fs.writeFileSync(path.join(dist, "buildInfo.json"), JSON.stringify({
            gitSha: null,
            builtAt: new Date().toISOString(),
            srcHash,
        }));
        const summary = getDevBuildSummary(tmp);
        expect(summary).not.toBeNull();
        expect(summary?.srcMatches).toBe(true);
    });
    it("reports srcMatches=false when src has changed since build", () => {
        const src = path.join(tmp, "src");
        const dist = path.join(tmp, "dist");
        fs.mkdirSync(src, { recursive: true });
        fs.mkdirSync(dist, { recursive: true });
        fs.writeFileSync(path.join(src, "a.ts"), "export const x = 1\n");
        const staleHash = hashSrcDir(src);
        fs.writeFileSync(path.join(src, "a.ts"), "export const x = 2\n");
        fs.writeFileSync(path.join(dist, "buildInfo.json"), JSON.stringify({
            gitSha: "abc1234",
            builtAt: new Date().toISOString(),
            srcHash: staleHash,
        }));
        const summary = getDevBuildSummary(tmp);
        expect(summary?.srcMatches).toBe(false);
    });
});
describe("formatRelativeTime", () => {
    it("formats seconds, minutes, hours, days", () => {
        const now = new Date("2026-04-15T12:00:00Z").getTime();
        expect(formatRelativeTime("2026-04-15T11:59:30Z", now)).toBe("30s ago");
        expect(formatRelativeTime("2026-04-15T11:55:00Z", now)).toBe("5m ago");
        expect(formatRelativeTime("2026-04-15T09:00:00Z", now)).toBe("3h ago");
        expect(formatRelativeTime("2026-04-13T12:00:00Z", now)).toBe("2d ago");
    });
});
describe("formatters", () => {
    const summary = {
        info: {
            gitSha: "abc12345",
            builtAt: new Date().toISOString(),
            srcHash: "deadbeef0000",
        },
        pluginRoot: "/path/to/plugin",
        srcMatches: true,
        relativeBuiltAt: "2m ago",
    };
    it("status line shows short sha + match indicator", () => {
        expect(formatDevBuildStatusLine(summary)).toBe("Build: abc1234 · 2m ago · src matches: ✅");
    });
    it("banner lines include plugin path and only warn on mismatch", () => {
        expect(formatDevBuildBannerLines(summary, "Bitfab")).toEqual([
            "[Bitfab Dev] build abc1234 · 2m ago · /path/to/plugin",
        ]);
        expect(formatDevBuildBannerLines({ ...summary, srcMatches: false }, "Bitfab")).toEqual([
            "[Bitfab Dev] build abc1234 · 2m ago · /path/to/plugin",
            "             src changed since last build — run pnpm build to apply.",
        ]);
    });
});
