// bitfab-plugin-lib/dist/handoffTickets.js
const DEFAULT_POLL_INTERVAL_MS = 1500;
export async function createHandoffTicket(serviceUrl, kind, opts) {
    const headers = {
        "Content-Type": "application/json",
    };
    if (opts?.bearer) {
        headers.Authorization = `Bearer ${opts.bearer}`;
    }
    const res = await fetch(`${serviceUrl}/api/plugin/handoff-tickets`, {
        method: "POST",
        headers,
        body: JSON.stringify({
            kind,
            pkceChallenge: opts?.pkceChallenge,
            pkceMethod: opts?.pkceMethod,
            ttlSeconds: opts?.ttlSeconds,
        }),
    });
    if (!res.ok) {
        const body = await res.json().catch(() => ({}));
        throw new Error(body.error ??
            `Failed to create handoff ticket (${res.status})`);
    }
    return (await res.json());
}
export async function pollHandoffTicket(serviceUrl, ticketId, opts) {
    const interval = opts.intervalMs ?? DEFAULT_POLL_INTERVAL_MS;
    const deadline = Date.now() + opts.timeoutMs;
    while (Date.now() < deadline) {
        if (opts.abortSignal?.aborted) {
            return { expired: true };
        }
        const url = new URL(`${serviceUrl}/api/plugin/handoff-tickets/${ticketId}`);
        if (opts.verifier) {
            url.searchParams.set("verifier", opts.verifier);
        }
        const headers = {};
        if (opts.bearer) {
            headers.Authorization = `Bearer ${opts.bearer}`;
        }
        try {
            const res = await fetch(url.toString(), { headers });
            if (res.ok) {
                const body = (await res.json());
                if (body.status === "completed" && body.data) {
                    return { data: body.data };
                }
                if (body.status === "expired" || body.status === "cancelled") {
                    return { expired: true };
                }
            }
        }
        catch {
            // Network error — retry on next interval
        }
        await new Promise((resolve, reject) => {
            const timer = setTimeout(resolve, interval);
            opts.abortSignal?.addEventListener("abort", () => {
                clearTimeout(timer);
                reject(new DOMException("Aborted", "AbortError"));
            }, { once: true });
        }).catch(() => { });
    }
    return { expired: true };
}
