// bitfab-plugin-lib/dist/handoffTicketChannel.js
import { createHandoffTicket, pollHandoffTicket } from "./handoffTickets.js";
import { generatePkce } from "./pkce.js";
/**
 * Parallel abstraction to `startLoopbackServer`: a single self-contained
 * delivery path that uses a server-polled handoff ticket with PKCE instead
 * of a local HTTP callback. Works in environments where the browser can't
 * reach the CLI's loopback (SSH, Docker, WSL, cloud IDEs, sandboxes) because
 * the hand-off goes browser → server → CLI poll.
 *
 * Creates the ticket, kicks off polling in the background, and exposes the
 * resulting promise. The caller races this against the loopback server's
 * promise — typically via `runBrowserHandoff` in `browserHandoff.ts`.
 *
 * The generic `T` matches the shape of the ticket's `data` payload as
 * validated by the server (see `handoffTicket.ts` domain `dataSchema`). For
 * example, `auth.login` tickets hydrate to `{ token: string }`;
 * `dataset.start` tickets pass through `{ status, isFetch, ... }`.
 */
export async function startHandoffTicketChannel(opts) {
    const { verifier, challenge } = generatePkce();
    const abort = new AbortController();
    let done = false;
    let resolveFn = () => { };
    let rejectFn = () => { };
    const promise = new Promise((resolve, reject) => {
        resolveFn = (value) => {
            if (done) {
                return;
            }
            done = true;
            resolve(value);
        };
        rejectFn = (err) => {
            if (done) {
                return;
            }
            done = true;
            reject(err);
        };
    });
    let startError = null;
    let ticket = null;
    try {
        ticket = await createHandoffTicket(opts.serviceUrl, opts.kind, {
            pkceChallenge: challenge,
            pkceMethod: "S256",
            bearer: opts.bearer,
            ttlSeconds: opts.ttlSeconds,
        });
    }
    catch (err) {
        startError = err instanceof Error ? err : new Error(String(err));
    }
    const available = ticket !== null;
    if (ticket) {
        pollHandoffTicket(opts.serviceUrl, ticket.id, {
            verifier,
            bearer: opts.bearer,
            timeoutMs: opts.timeoutMs,
            intervalMs: ticket.pollIntervalMs,
            abortSignal: abort.signal,
        })
            .then((result) => {
            if ("expired" in result) {
                rejectFn(new Error("Handoff ticket expired before completion."));
                return;
            }
            resolveFn(result.data);
        })
            .catch((err) => {
            rejectFn(err instanceof Error ? err : new Error(String(err)));
        });
    }
    else {
        // The channel is unusable but callers still have a handle to `promise`
        // and `reject`. Silently absorb any rejection so a `reject()` call from
        // the caller doesn't surface as an unhandled promise rejection.
        promise.catch(() => { });
    }
    return {
        promise,
        get done() {
            return done;
        },
        available,
        startError,
        ticketId: ticket?.id ?? null,
        reject: rejectFn,
        close: () => {
            abort.abort();
            rejectFn(new Error("Handoff ticket channel closed."));
        },
    };
}
